mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergOutputFormat.java

```diff
@@ -19,21 +19,442 @@
+   * @param conf The job's configuration
+   * @param taskAttemptId The TaskAttemptID for the task
+   * @return The file to store the results
+   */
+  private static String generateCommitFileLocation(Configuration conf, TaskAttemptID taskAttemptId) {
+    return generateCommitFileLocation(conf, taskAttemptId.getJobID(), taskAttemptId.getTaskID().getId());
+  }
+
+  @Override
+  public org.apache.hadoop.mapred.RecordWriter<NullWritable, IcebergWritable> getRecordWriter(FileSystem ignored,
+      JobConf job, String name, Progressable progress) {
+
+    throw new UnsupportedOperationException("Please implement if needed");
   }
 
   @Override
   public void checkOutputSpecs(FileSystem ignored, JobConf job) {
-    throw new UnsupportedOperationException("Writing to an Iceberg table with Hive is not supported");
+    // Not doing any check.
+  }
+
+  protected class IcebergRecordWriter extends org.apache.hadoop.mapreduce.RecordWriter<NullWritable, IcebergWritable>
+      implements FileSinkOperator.RecordWriter, org.apache.hadoop.mapred.RecordWriter<NullWritable, IcebergWritable> {
+
+    private final String location;
+    private final FileAppender<Record> appender;
+    private final FileIO io;
+
+    IcebergRecordWriter(String location) throws IOException {
+      this.location = location;
+      io = new HadoopFileIO(overlayedConf);
+      OutputFile dataFile = io.newOutputFile(location);
+
+      switch (fileFormat) {
+        case AVRO:
+          this.appender = Avro.write(dataFile)
+              .schema(schema)
+              .createWriterFunc(DataWriter::create)
+              .named(fileFormat.name())
+              .build();
+          break;
+
+        case PARQUET:
+          this.appender = Parquet.write(dataFile)
+              .schema(schema)
+              .createWriterFunc(GenericParquetWriter::buildWriter)
+              .named(fileFormat.name())
+              .build();
+          break;
+
+        case ORC:
+          this.appender = ORC.write(dataFile)
+              .schema(schema)
+              .createWriterFunc(GenericOrcWriter::buildWriter)
+              .build();
+          break;
+
+        default:
+          throw new UnsupportedOperationException("Cannot write format: " + fileFormat);
+      }
+      LOG.info("IcebergRecordWriter is created in {} with {}", location, fileFormat);
+    }
+
+    @Override
+    public void write(Writable row) {
+      Preconditions.checkArgument(row instanceof IcebergWritable);
+
+      // TODO partition handling
+      appender.add(((IcebergWritable) row).record());
+    }
+
+    @Override
+    public void write(NullWritable key, IcebergWritable value) {
+      write(value);
+    }
+
+    @Override
+    public void close(boolean abort) throws IOException {
+      appender.close();
+      if (!abort) {
+        fileData.put(taskAttemptId, new ClosedFileData(location, fileFormat, appender.length(), appender.metrics()));
```

Comment:
Just checking, are we sure it's OK to call `appender.length()` after `close()` has been called? There aren't any cases where a closed appender would throw an exception if actions were performed on it after close?

Reply:
Good question!
Checked now, and here is what I have found:
- metrics() - "Only valid after the file is closed" states the javadoc comment, so I think we are good here
- length() - I have found several other uses for it (BaseTaskWriter.closeCurrent(), EqualityDeleteWriter.close()), so if I do not miss something obvious (or less obvious 😄) then it is ok to use it after close.

Reply:
OK, cool, thanks for double checking.

Reply:
Yes, `appender.length()` is intended to be valid after the file is closed.

Reply:
Thanks for the confirmation @rdblue!